File: main/java/com/websarva/wings/android/dowasurememo/AddressMemo.java
package com.websarva.wings.android.dowasurememo;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;

/**
 *住所を入力するAddressメモのクラス
 * @author nakayama
 * @version 1.0.2
 */
public class AddressMemo extends AppCompatActivity {

    private Databasehelper _helper;
    private static final String _CATEGORY = "ADDRESS";

    /**
     *データベースに保存する際のID値
     */
    private int tagId=2;

    /**
     * データベースのテーブル名
     */
    private static final String TABLE="address";

    private Context context=AddressMemo.this;

    private LayoutInflater inflater;
    private LinearLayout llAddressLayout;
    private LinearLayout llAddressInputform;
    private LinearLayout llAddressFrame;
    private LinearLayout llPostNumberinputform;

    private EditText etAddressTitle;
    private EditText etPostNumber1;
    private EditText etPostNumber2;
    private EditText etAddressDetail;
    private ImageButton btDelete;

    private String strAddressTitle;
    private String strPostNumber1;
    private String strPostNumber2;
    private String strAddressDetail;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_address_memo);

        Intent intent=getIntent();
        inflater=LayoutInflater.from(getApplicationContext());
        llAddressLayout=findViewById(R.id.ll_address_layout);
        llAddressInputform= (LinearLayout) inflater.inflate(R.layout.address_inputform,null);

//        llAddressLayout.removeAllViews();
//        DatabaseControl control4=new DatabaseControl(context,table);
//        control4.DatabaseAllDelete();

        /**
         * データベースの列名の配列
         */
        String[] columnNames={"addresstitle","postnumber1","postnumber2","addressdetail"};

        /**
         * データベースからデータを取り出して、レイアウトを作成する
         * @param context　コンテキスト
         * @param TABLE　テーブル名
         * @param columnNames　データベースの列名の配列
         */
        DatabaseControl control=new DatabaseControl(context,TABLE,columnNames);
        control.selectDatabase(llAddressLayout,llAddressInputform);

    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater=getMenuInflater();
        inflater.inflate(R.menu.optionmenu,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.option_add:
                inflater=LayoutInflater.from(getApplicationContext());
                llAddressLayout=findViewById(R.id.ll_address_layout);
                llAddressInputform = (LinearLayout) inflater.inflate(R.layout.address_inputform, null);
                llAddressLayout.addView(llAddressInputform);

                llAddressFrame=llAddressInputform.findViewById(R.id.ll_address_frame);
                llPostNumberinputform=llAddressFrame.findViewById(R.id.ll_postnumber_inputform);
                btDelete=llPostNumberinputform.findViewById(R.id.bt_delete);
                btDelete.setOnClickListener
                        (new DeleteButton(AddressMemo.this,llAddressLayout,llAddressInputform,TABLE));

        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onPause() {
        super.onPause();

        tagId = 2;

        for (int i = 0; i < llAddressLayout.getChildCount(); i++) {
            LinearLayout linearLayout = (LinearLayout) llAddressLayout.getChildAt(i);
            etAddressTitle = linearLayout.findViewById(R.id.et_address_title);
            etPostNumber1 = linearLayout.findViewById(R.id.et_postnumber1);
            etPostNumber2 = linearLayout.findViewById(R.id.et_postnumber2);
            etAddressDetail = linearLayout.findViewById(R.id.et_addres_detail);

            strAddressTitle = etAddressTitle.getText().toString();
            strPostNumber1 = etPostNumber1.getText().toString();
            strPostNumber2 = etPostNumber2.getText().toString();
            strAddressDetail = etAddressDetail.getText().toString();

            DatabaseControl control = new DatabaseControl(context, TABLE);
            control.deleteDatabase(tagId);

            String column1="addresstitle";
            String column2="postnumber1";
            String column3="postnumber2";
            String column4="addressdetail";

            DatabaseControl control2 = new DatabaseControl
                    (context, TABLE, tagId, _CATEGORY, strAddressTitle,strPostNumber1,strPostNumber2,strAddressDetail);
            control2.insertDatabaseFourColumns(column1, column2, column3, column4);

            tagId++;

        }
    }

}

File: main/java/com/websarva/wings/android/dowasurememo/AddressMemo.kt
package com.websarva.wings.android.dowasurememo

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.widget.*
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.websarva.wings.android.dowasurememo.databinding.ActivityAddressMemoBinding
import com.websarva.wings.android.dowasurememo.databinding.AddressInputformBinding

/**
 * 住所を入力するAddressメモのクラス
 *
 * @author nakayama
 * @version 1.0
 */
class AddressMemo : AppCompatActivity() {

    companion object {
        private const val _CATEGORY = "ADDRESS"

        /**
         * データベースのテーブル名
         */
        private const val TABLE = "address"
    }

    private val context: Context = this@AddressMemo
    private lateinit var binding: ActivityAddressMemoBinding
    private lateinit var inputformBinding: AddressInputformBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddressMemoBinding.inflate(layoutInflater)
        inputformBinding = AddressInputformBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val llAddressLayout = findViewById<LinearLayout>(R.id.ll_address_layout)

        /**
         * データベースの列名の配列
         */
        val columnNames = listOf("addresstitle", "postnumber1", "postnumber2", "addressdetail")

        //データベースからデータを取り出して、レイアウトを作成する処理
        DatabaseControl(context, TABLE, columnNames)
                .selectDatabase(llAddressLayout)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.optionmenu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.option_add -> {
                //オプションメニューの＋ボタンを押すと、動的にビューを追加する処理
                val inflater = LayoutInflater.from(applicationContext)
                val llAddressLayout = findViewById<LinearLayout>(R.id.ll_address_layout)
                val llAddressInputform = inflater.inflate(R.layout.address_inputform, null) as LinearLayout
                val llAddressFrame = llAddressInputform.findViewById<LinearLayout>(R.id.ll_address_frame)
                val llAddressTitle = llAddressFrame.findViewById<LinearLayout>(R.id.ll_address_title)
                val btDelete = llAddressTitle.findViewById<ImageButton>(R.id.bt_delete)
                btDelete.setOnClickListener {
                    DeleteButton(this@AddressMemo, llAddressLayout, llAddressInputform, TABLE)
                }

                val llPostNumberinputform = llAddressFrame.findViewById<LinearLayout>(R.id.ll_postnumber_inputform)
                val btPostNumberSearch = llPostNumberinputform.findViewById<Button>(R.id.bt_postnumber_search)
                val etPostNumber1 = llPostNumberinputform.findViewById<EditText>(R.id.et_postnumber1)
                val etPostNumber2 = llPostNumberinputform.findViewById<EditText>(R.id.et_postnumber2)
                val etAddressDetail = llAddressFrame.findViewById<EditText>(R.id.et_address_detail)
                btPostNumberSearch.setOnClickListener {
                    val addressText = PostNumberAPIClient(this)
                            .receivePostNumberInfo(etPostNumber1.text.toString(), etPostNumber2.text.toString())
                    Log.d("button", "$addressText")
                    etAddressDetail.setText(addressText)
                }
                llAddressLayout.addView(llAddressInputform)
            }
        }
        return super.onOptionsItemSelected(item)
    }


    override fun onPause() {
        super.onPause()

        //データベースにある全てのデータを削除
        DatabaseControl(context, TABLE).deleteAllDatabase()

        //メモの文字列を取得してデータベースにインサートする
        for (i in 0..binding.llAddressLayout.childCount - 1) {
            val linearLayout = binding.llAddressLayout.getChildAt(i) as LinearLayout

            val etAddressTitle = linearLayout.findViewById<EditText>(R.id.et_address_title)
            val etPostNumber1 = linearLayout.findViewById<EditText>(R.id.et_postnumber1)
            val etPostNumber2 = linearLayout.findViewById<EditText>(R.id.et_postnumber2)
            val etAddressDetail = linearLayout.findViewById<EditText>(R.id.et_address_detail)

            val strAddressTitle = etAddressTitle.getText().toString()
            val strPostNumber1 = etPostNumber1.getText().toString()
            val strPostNumber2 = etPostNumber2.getText().toString()
            val strAddressDetail = etAddressDetail.getText().toString()
            val control2 = DatabaseControl(context, TABLE, i, _CATEGORY, strAddressTitle, strPostNumber1, strPostNumber2, strAddressDetail)
            control2.insertDatabaseFourColumns("addresstitle", "postnumber1", "postnumber2", "addressdetail")
        }
        finish()
    }


}


